qiskit/pulse/filters.py

```diff
@@ -17,48 +17,86 @@
 
 import numpy as np
 
-from qiskit.pulse import Schedule
+from qiskit.pulse import Schedule, ScheduleBlock, Instruction
 from qiskit.pulse.channels import Channel
 from qiskit.pulse.schedule import Interval
+from qiskit.pulse.exceptions import PulseError
 
 
 def filter_instructions(
-    sched: Schedule, filters: List[Callable], negate: bool = False, recurse_subroutines: bool = True
-) -> Schedule:
+    sched: Union[Schedule, ScheduleBlock],
+    filters: List[Callable],
+    negate: bool = False,
+    recurse_subroutines: bool = True,
+) -> Union[Schedule, ScheduleBlock]:
     """A filtering function that takes a schedule and returns a schedule consisting of
     filtered instructions.
 
     Args:
-        sched: A pulse schedule to be filtered.
+        sched: A pulse schedule or schedule_block to be filtered.
         filters: List of callback functions that take an instruction and return boolean.
         negate: Set `True` to accept an instruction if a filter function returns `False`.
             Otherwise the instruction is accepted when the filter function returns `False`.
         recurse_subroutines: Set `True` to individually filter instructions inside of a subroutine
             defined by the :py:class:`~qiskit.pulse.instructions.Call` instruction.
 
     Returns:
-        Filtered pulse schedule.
+        Filtered pulse schedule or schedule_block.
     """
     from qiskit.pulse.transforms import flatten, inline_subroutines
 
-    target_sched = flatten(sched)
-    if recurse_subroutines:
-        target_sched = inline_subroutines(target_sched)
+    if isinstance(sched, Schedule):
```

Comment:
You can use singledispatchmethod if you want, e.g. https://github.com/Qiskit/qiskit-terra/pull/8944/files#diff-68d7c1d0fc4ee6a37bac4136c7c7ee93faf85f8cbcd1bb3dc7c21165249476a6

Reply:
Thank you! I never knew it. I use it.